### PR TITLE
[SPARK-9018][MLLIB] add stopwatches

### PR DESCRIPTION
**Pull request by mengxr**
Add stopwatches for easy instrumentation of MLlib algorithms. This is based on the `TimeTracker` used in decision trees. The distributed version uses Spark accumulator. @jkbradley

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/util/stopwatches.scala [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    startTime = now
+  }
+
+  /**
+   * Stops the stopwatch and returns the duration of the last session in milliseconds.
+   * Throws an exception if the stopwatch is not running.
+   */
+  def stop(): Long = {
+    assume(running, "stop() called but the stopwatch is not running.")
+    val duration = now  - startTime
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/stopwatches.scala]: extra space!

File: mllib/src/main/scala/org/apache/spark/ml/util/stopwatches.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+/**
+ * A distributed [[Stopwatch]] using Spark accumulator.
+ * @param sc SparkContext
+ */
+private[spark] class DistributedStopwatch(
+    sc: SparkContext,
+    override val name: String) extends Stopwatch {
+
+  val elapsedTime: Accumulator[Long] = sc.accumulator(0L, s"DistributedStopwatch($name)")
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/stopwatches.scala]: private

File: mllib/src/main/scala/org/apache/spark/ml/util/stopwatches.scala [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  /**
+   * Gets a stopwatch.
+   * @param name stopwatch name
+   */
+  def apply(name: String): Stopwatch = stopwatches(name)
+
+  override def toString: String = {
+    "MultiStopwatch" +
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/stopwatches.scala]: I don't think this should print a title.  If I used it to time an algorithm, I'd prefer it read:
```
DecisionTreeClassifier(
  init -> 17ms,
  ...
)
```

File: mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    assert(!sw.isRunning)
+    intercept[AssertionError] {
+      sw.stop()
+    }
+    sw.start()
+    Thread.sleep(50)
+    val duration = sw.stop()
+    assert(duration >= 50 && duration < 100) // using a loose upper bound
+    val elapsed = sw.elapsed()
+    assert(elapsed >= 50 && elapsed < 100)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala]: Why not check for strict equality with "duration?"

File: mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    sw.start()
+    Thread.sleep(50)
+    val duration2 = sw.stop()
+    assert(duration2 >= 50 && duration2 < 100)
+    val elapsed2 = sw.elapsed()
+    assert(elapsed2 >= 100 && elapsed2 < 200)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala]: check for equality with duration + duration2

**Comment by jkbradley** [comment]
That's all I see.  But also note Sean's comments in the JIRA.

File: mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    val duration = sw.stop()
+    assert(duration >= 50 && duration < 100) // using a loose upper bound
+    val elapsed = sw.elapsed()
+    assert(elapsed === duration)
+    sw.start()
+    Thread.sleep(50)
+    val duration2 = sw.stop()
+    assert(duration2 >= 50 && duration2 < 100)
+    val elapsed2 = sw.elapsed()
+    assert(elapsed2 == duration + duration2)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala]: triple equals

**Comment by jkbradley** [comment]
LGTM pending tests

**Comment by jkbradley** [comment]
I'll let you fix that, or merge, as you like

File: mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala (hunk 6 above)
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala]: Actually @ericl pointed out `==` and `===` are equal in this case (Long). Both provide the same error message. I will update it to be consistent.

**Comment by mengxr** [comment]
Merged into master.

File: mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala (hunk 6 above)
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/util/StopwatchSuite.scala]: Should we no longer bother with this?  Or is it just for Longs (in which case enforcing consistency may be easiest)?

**Pull request closed without merging**